aws/rust-runtime/aws-http/src/request_id.rs

```diff
@@ -0,0 +1,181 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+use aws_smithy_http::http::HttpHeaders;
+use aws_smithy_http::operation;
+use aws_smithy_http::result::SdkError;
+use aws_smithy_types::error::{
+    Builder as GenericErrorBuilder, Error as GenericError, ErrorMetadata, Unhandled,
+};
+use http::{HeaderMap, HeaderValue};
+
+/// Constant for the [`aws_smithy_types::error::Error`] extra field that contains the request ID
+const AWS_REQUEST_ID: &str = "aws_request_id";
+
+/// Implementers add a function to return an AWS request ID
+pub trait RequestId {
+    /// Returns the request ID if it's available.
+    fn request_id(&self) -> Option<&str>;
+}
+
+impl<E, R> RequestId for SdkError<E, R>
+where
+    R: HttpHeaders,
+{
+    fn request_id(&self) -> Option<&str> {
+        match self {
+            Self::ResponseError(err) => extract_request_id(err.raw().http_headers()),
+            Self::ServiceError(err) => extract_request_id(err.raw().http_headers()),
```

Comment:
Is this valid?
```suggestion
            Self::ResponseError(err) | Self::ServiceError(err) => extract_request_id(err.raw().http_headers()),
```

Reply:
This only works if `err` is the same type between the two, which it is not in this case.